File: jlpt/native_android/app/src/main/java/com/jlpt/android/domain/util/AppResult.kt
package com.jlpt.android.domain.util

/** Outcome of a repository/use-case call, carrying a human-readable message on failure. */
sealed class AppResult<out T> {
    data class Success<T>(val data: T) : AppResult<T>()
    data class Failure(val message: String, val cause: Throwable? = null) : AppResult<Nothing>()
}

inline fun <T, R> AppResult<T>.map(transform: (T) -> R): AppResult<R> = when (this) {
    is AppResult.Success -> AppResult.Success(transform(data))
    is AppResult.Failure -> this
}

inline fun <T> AppResult<T>.onSuccess(action: (T) -> Unit): AppResult<T> {
    if (this is AppResult.Success) action(data)
    return this
}

inline fun <T> AppResult<T>.onFailure(action: (String) -> Unit): AppResult<T> {
    if (this is AppResult.Failure) action(message)
    return this
}

File: jlpt/native_ios/Sources/Domain/Util/AppResult.swift
import Foundation

/// Outcome of a repository/use-case call, carrying a human-readable message on failure.
enum AppResult<T> {
    case success(T)
    case failure(message: String, cause: Error? = nil)

    func map<R>(_ transform: (T) -> R) -> AppResult<R> {
        switch self {
        case .success(let value): return .success(transform(value))
        case .failure(let message, let cause): return .failure(message: message, cause: cause)
        }
    }

    @discardableResult
    func onSuccess(_ action: (T) -> Void) -> AppResult<T> {
        if case .success(let value) = self { action(value) }
        return self
    }

    @discardableResult
    func onFailure(_ action: (String) -> Void) -> AppResult<T> {
        if case .failure(let message, _) = self { action(message) }
        return self
    }
}
